### ISSUE TITLE
Introduce optimization cache

### ISSUE THREAD

**Issue by kelson42** [openzim/gutenberg#101]
Like in MWoffliner, see https://github.com/openzim/mwoffliner/issues/996

**Comment by satyamtg** [comment]
I tried doing it today. However, it seems that it's a bit tricky with how Gutenberg is currently structured. I think it may be a better idea to zip all files related to a book and then upload it due to the following -
- We don't really know whether a book is zipped or not (i.e. has HTML available directly) before iterating over the possible URLs
- We don't have a system in place to know what files are related to a book (maybe inside a zip) before actually having it downloaded.
I think it's better that we discuss the pros and cons before proceeding as I currently do not have a very solid idea on how much the performance will be impacted if we zip (maybe without compression) and upload the optimized files.

**Comment by satyamtg** [comment]
Okay, so after messing up with gutenberg and aleph.gutenberg.org , I can tell the following things - 
- the setup_urls() is a very important step if we want to implement the s3 cache without zipping up files. That said, currently, the scraper downloads the files as follows -

- retrieve the book format from the database
- retrieve the possible URLs from the database (if setup_urs() is not run, it'll send all possible combinations)
- download a zip or a file from the URL (which is done in urls.py) or else show error
- if download was a zip, we extract it and use its contents to form HTML (the optimized one
And here's the problem. Although the scraper tries to retrieve a full URL list, it doesn't seem to do so. I mean, we have the HTML files hosted also on the aleph.gutenberg.org server, the individual images and contents of the zip. So, if we want to know what files belong to which book, before downloading them (for the s3 usage), we actually can do that (as we did the rsync). But we currently don't do that. We use files extracted from the zip.

Also, the downloaded zips are never deleted  from the download cache directory and take a lot of space there.

So, here's the question - 
Should we go with downloading files which are already unzipped directly? If so, the con would be many more requests to the server, but we can, of course, parallelize (to an extent)

IMO, a better option to avoid this would be to actually have different methods to get URLs for both the S3 cache and the main downloading function (the one which downloads from the site). However, this would probably make the complex code a bit more complexier.

For uploading to S3 cache, I don't think its a problem if we know what files we download. But all this is only possible with the rsync step (which takes quite long)

@rgaudin @dattaz @kelson42 what are your views ?

**Comment by rgaudin** [comment]
Thank you very much for taking the time to dig into this; I know it’s not easy.

I did not quite understand how it works from your summary. Can you elaborate a bit more, assuming we don’t know the source code (it’s been years!)

- what does setup_urls does exactly?
- what are the book format info from the DB you mention? where does it come from ? the RDF? 
- urls retrieval : what’s the two possible state of the DB and what does each method involve? Is that where we hit all urls or are those just possible links used later on?
- is it right to assume there is only one target for each book ? a zip or another file ?
- I didn’t get what’s in the last billet point...

Sorry about all the questions but it’s still too fuzzy for me to respond at the moment.

As for the files not being deleted I believe it was on purpose as it was a pain to retrieve but yes there should be an option to keep them and otherwise remove it all.

**Comment by satyamtg** [comment]
> * what does setup_urls does exactly?

Okay, to be exact, it just gets the list of all files on the mirrors (say aleph.gutenberg.org), which is of this form -
```
drwsrwsr-x          4,096 2015/04/14 22:51:05 0
drwxrwxr-x          4,096 2018/08/22 19:38:40 0/1
-rw-rw-r--        120,941 2018/08/12 02:25:58 0/1/1-0.txt
-rw-rw-r--         45,685 2018/08/12 02:36:48 0/1/1-0.zip
-rw-rw-r--      2,189,095 2018/08/12 02:36:48 0/1/1-h.zip
drwxrwxr-x          4,096 2018/08/12 01:50:20 0/1/1-h
-rw-rw-r--        131,866 2018/08/12 02:35:44 0/1/1-h/1-h.htm
drwxrwxr-x          4,096 2018/08/12 01:50:20 0/1/1-h/aimages
-rw-rw-r--        398,827 2018/08/12 01:34:52 0/1/1-h/aimages/dec1.jpg
-rw-rw-r--        124,596 2018/08/12 01:34:52 0/1/1-h/aimages/dec1th.jpg
-rw-rw-r--        410,127 2018/08/12 01:34:52 0/1/1-h/aimages/dec2.jpg
-rw-rw-r--        127,220 2018/08/12 01:34:52 0/1/1-h/aimages/dec2th.jpg
-rw-rw-r--        407,770 2018/08/12 01:34:52 0/1/1-h/aimages/dec3.jpg
-rw-rw-r--        130,929 2018/08/12 01:34:52 0/1/1-h/aimages/dec3th.jpg
-rw-rw-r--        410,335 2018/08/12 01:34:52 0/1/1-h/aimages/dec4.jpg
-rw-rw-r--        131,909 2018/08/12 01:34:52 0/1/1-h/aimages/dec4th.jpg
```
gets only the relative URLs, i.e. -
```
0
0/1
0/1/1-0.txt
0/1/1-0.zip
0/1/1-h.zip
```
and inserts it into the database in the urls column.

**Comment by rgaudin** [comment]
Thank you **very much** for this detailed description. It's all a lot more clear now (and it allowed me to discover and fix a small issue).

Now, I know that the rsync step was added later-on because the other method wasn't solid enough.
Given the fact that we're always downloading from that same mirror (according to `UrlBuilder`), I think it's safe to assume that if we can't find the link in the DB, it won't be downloadable and thus considered missing.

So I'd suggest we remove the former code and use only the rsync/db one. I might require some refinement/refactoring as the two were mixed together from my understanding.

With this secured, I think it will be easy to implement the S3 cache. I see that the server supports `ETag` which we could use as a version meta on our storage to check whether it had been updated or not.

@satyamtg what do you think?

**Comment by satyamtg** [comment]
We have a couple of options now that we know server has both zips and the extracted files -
Method 1
- Use extracted files for normal downloads from server
- Use extracted files for normal downloads from S3 cache
Pros -
- Simple to implement
- No zipping required at time of uploading to S3
Cons -
- Code needs to be refactored to support the new data from DB
- Many more requests per book

Method 2
- Use ZIP for normal downloads from server
- Use extracted files for normal downloads from S3 cache
Pros -
- No zipping required at time of uploading to S3
- Fewer HTTP requests to server in comparison to Method 1
Cons -
- Code needs to be refactored to support the new data from DB

Method 3
- Use ZIP for normal downloads from server
- Use ZIP for normal downloads from S3 cache
Pros -
- No code refactoring required for getting new data
Cons -
- Zipping required, which increases computation overhead

I find going with 1 or 2 more logical but we need to make an informed choice. What do you think @kelson42 @rgaudin

**Comment by rgaudin** [comment]
Thank you @satyamtg but I think it's hardly understandable by those without the full context 😅 Let me rephrase:

* We are dealing with three formats for books: `epub`, `pdf` and `html`.
* `epub` and `pdf` are single files on gut server so it's a no brainer.
* `html` is usually available in two different forms: zipped content (so a single file) and exploded content (multiple files).
* It's not clear to me whether **all books** with HTML have both or if some have only a single form. This should be verifiable (and verified) with the database.
* we download epub files and produce optimized epub files.
* we download pdf file and don't optimize them. we could/should though at some point.
* we download HTML based on what's available with a priority on those with images (ZIP thus) so we may sometimes get a ZIP and sometimes get a single HTML file. Again, we could check that from the database to see how frequent that is.
* we extract the ZIP content and optimize everything in it but we don't produce a new ZIP file as we use it exploded in the ZIM.
* we optimize the single HTML as well.

That was for the context. The question is, regarding those HTML files, should we:

1. skip the ZIP and treat each file individually (download each html and image files one by one,  optimize them one by one and upload/retrieve one by one from S3)
1. download ZIP from gut but store individual files in S3
1. consider the ZIP an entity and download it, optimize it, then re-archive it to zip (new) and store that in S3.

The pros/cons revolve around chosing what we want to save:

* gutenberg requests? If so, shouldn't download as indiv.
* S3 network requests? If so, should store HTML as zip.
* CPU cycles? If so, should not zip back.

I think my opinion leans towards solution 3: using ZIP from gut and storing optimized ZIP. It's network efficient, it's *clean* in that we share a single 1:1 concept between what's downloaded and what's stored and as for CPU, I think uncompressed ZIP or fast-compressed ZIP are not a problem.

_Note_: our _optimization_ goes beyond saving space. We strip some content from the file (gutenberg banners at the beginning of files for instance) and we rewrite the links so that they can work from inside the ZIM.

**Comment by satyamtg** [comment]
Okay, going with 3 seems a great option. Also it would increase the CPU cycles, but that shouldn't be a problem with uncompressed ZIPs. Moreover, this overhead would be in the first run(s) only.

**Comment by eshellman** [comment]
Sorry I've been working on something else, but I can answer questions.

The complete file list for each book is in the rdf files; it might be avoid a lot of flailing if you just used that. Also, while I love BS4, it's not always an efficient way to parse rdf. I can see if I can insert my code if you like.
